File: backend/src/app/seed.py
from __future__ import annotations

import json
import logging
from datetime import datetime
from decimal import Decimal
from pathlib import Path

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.worklogs.models import Adjustment, TimeSegment, Worklog
from app.worklogs.validation import (
    validate_amount,
    validate_non_negative_hourly_rate,
    validate_segment_interval,
)

logger = logging.getLogger(__name__)


def _parse_dt(value: str) -> datetime:
    return datetime.fromisoformat(value.replace("Z", "+00:00"))


def load_seed_json(session: Session, path: str) -> None:
    """Load assessment seed file into empty tables (idempotent if worklogs already exist)."""
    file_path = Path(path)
    if not file_path.is_file():
        logger.warning("Seed file not found at %s; skipping seed.", path)
        return

    existing = session.scalar(select(func.count()).select_from(Worklog))
    if existing is not None and existing > 0:
        return

    rows = json.loads(file_path.read_text(encoding="utf-8"))
    for row in rows:
        worklog = Worklog(
            external_id=row["worklog_id"],
            user_id=row["user_id"],
            user_name=row["user_name"],
            task_name=row["task_name"],
            hourly_rate=validate_non_negative_hourly_rate(row["hourly_rate"]),
        )
        session.add(worklog)
        session.flush()

        for segment in row["segments"]:
            start = _parse_dt(segment["start"])
            end = _parse_dt(segment["end"])
            validate_segment_interval(start, end)
            session.add(
                TimeSegment(
                    worklog_id=worklog.id,
                    external_id=segment["segment_id"],
                    start=start,
                    end=end,
                    status=segment["status"],
                    dispute_reason=segment.get("dispute_reason"),
                )
            )

        for adjustment in row["adjustments"]:
            session.add(
                Adjustment(
                    worklog_id=worklog.id,
                    external_id=adjustment["adjustment_id"],
                    amount=validate_amount(adjustment["amount"], field_name="adjustment.amount"),
                    reason=adjustment["reason"],
                    applied_at=_parse_dt(adjustment["applied_at"]),
                )
            )

    session.commit()

File: backend/src/database/seed.js
const fs = require('fs');
const path = require('path');
const db = require('./pool');

/**
 * Resolves the path to the seed file.
 * In Docker the seed dir is at /app/seed; locally it's at repo root.
 * @returns {string}
 */
function resolveSeedPath() {
  const candidates = [
    path.join(__dirname, '..', '..', '..', 'seed', 'worklogs.json'),  // local dev
    path.join('/app', 'seed', 'worklogs.json'),                        // docker
  ];
  for (const p of candidates) {
    if (fs.existsSync(p)) return p;
  }
  return null;
}

/**
 * Seeds the database with worklogs, segments, and adjustments from the JSON file.
 * Skips silently if data already exists or the seed file is missing.
 */
async function seed() {
  const existing = await db.query('SELECT COUNT(*)::int AS cnt FROM worklog');
  if (existing.rows[0].cnt > 0) {
    console.log('[seed] Data already present — skipping.');
    return;
  }

  const seedPath = resolveSeedPath();
  if (!seedPath) {
    console.warn('[seed] No seed file found — skipping.');
    return;
  }

  const raw = fs.readFileSync(seedPath, 'utf8');
  const worklogs = JSON.parse(raw);

  const client = await db.getClient();
  try {
    await client.query('BEGIN');

    for (const wl of worklogs) {
      await client.query(
        `INSERT INTO worklog (worklog_id, user_id, user_name, task_name, hourly_rate)
         VALUES ($1, $2, $3, $4, $5)
         ON CONFLICT (worklog_id) DO NOTHING`,
        [wl.worklog_id, wl.user_id, wl.user_name, wl.task_name, wl.hourly_rate]
      );

      for (const seg of wl.segments || []) {
        await client.query(
          `INSERT INTO time_segment (segment_id, worklog_id, start_time, end_time, status, dispute_reason)
           VALUES ($1, $2, $3, $4, $5, $6)
           ON CONFLICT (segment_id) DO NOTHING`,
          [seg.segment_id, wl.worklog_id, seg.start, seg.end, seg.status, seg.dispute_reason || null]
        );
      }

      for (const adj of wl.adjustments || []) {
        await client.query(
          `INSERT INTO adjustment (adjustment_id, worklog_id, amount, reason, applied_at)
           VALUES ($1, $2, $3, $4, $5)
           ON CONFLICT (adjustment_id) DO NOTHING`,
          [adj.adjustment_id, wl.worklog_id, adj.amount, adj.reason, adj.applied_at]
        );
      }
    }

    await client.query('COMMIT');
    console.log(`[seed] Loaded ${worklogs.length} worklogs.`);
  } catch (err) {
    await client.query('ROLLBACK');
    console.error('[seed] Failed:', err.message);
    throw err;
  } finally {
    client.release();
  }
}

module.exports = { seed };
